File: swift/624_MaxDistanceInArrays.swift
class Solution {
    func maxDistance(_ arrays: [[Int]]) -> Int {
        var low: Int = arrays[0][0], high: Int = arrays[0].last!, maxDist: Int = 0
        for i in 1..<arrays.count {
            var currArray = arrays[i]
            maxDist = findExtrema([currArray.last! - low, high - currArray[0], maxDist])
            low = findExtrema([currArray[0], low], false)
            high = findExtrema([currArray.last!, high])
        }
        return maxDist
    }
    
    func findExtrema(_ nums: [Int], _ isMax: Bool = true) -> Int {
        var high = -1000000, low = 1000000
        for n in nums {
            high = n > high ? n : high
            low = n < low ? n : low
        }
        return isMax ? high : low
    }
}

File: cpp/624_MaxDistanceInArrays.cpp
class Solution {
public:
    int maxDistance(vector<vector<int>>& arrays) {
        int high = arrays[0].back(), low = arrays[0][0], max_dist = 0;
        for (int i = 1; i < arrays.size(); ++i) {
            vector<int> curr_array = arrays[i];
            max_dist = max(max(high - curr_array[0], curr_array.back() - low), max_dist);
            low = min(curr_array[0], low);
            high = max(curr_array.back(), high);
        }
        return max_dist;
    }
};
